compiler/dialects/stream.py

```diff
@@ -0,0 +1,161 @@
+from collections.abc import Sequence
+from typing import Generic, TypeVar
+
+from xdsl.dialects.builtin import (
+    AffineMapAttr,
+    AnyShapedType,
+    ArrayAttr,
+    ContainerType,
+    StringAttr,
+)
+from xdsl.dialects.utils import AbstractYieldOperation
+from xdsl.ir import (
+    Attribute,
+    Dialect,
+    IsTerminator,
+    ParametrizedAttribute,
+    Region,
+    SSAValue,
+    TypeAttribute,
+)
+from xdsl.irdl import (
+    AttrSizedOperandSegments,
+    IRDLOperation,
+    ParameterDef,
+    irdl_attr_definition,
+    irdl_op_definition,
+    opt_prop_def,
+    prop_def,
+    region_def,
+    var_operand_def,
+    var_result_def,
+)
+
+_StreamTypeElement = TypeVar("_StreamTypeElement", bound=Attribute)
+
+
+@irdl_attr_definition
+class StreamType(
+    Generic[_StreamTypeElement],
+    ParametrizedAttribute,
+    TypeAttribute,
+    ContainerType[_StreamTypeElement],
+):
+    """
+    A stream type with value sementics.
+    A stream is defined by an element type, and is produced as the result of
+    an operation, or through a streaming region op.
+    Streams can only be read from, there is no distinction between readable/writeable streams.
```

Comment:
Something I find a bit odd is that the stream would throw away all the shape information, but I guess a similar thing happens inside a linalg generic?

Also, I assume a stream can only be read from once, since reads "pop" from the stream?
Should we also enforce single SSA use somewhere? (If so, how?)